### PR TITLE
Feature of 2pc transaction api

### PR DESCRIPTION
**Pull request by jackycchen**
Update List:
1. add 2pc code according to go-client logic.
2. add some unit test for 2pc transaction api

### CODE REVIEW

**Comment by birdstorm** [comment]
Can we move the `proto` files to its original places(root directory) so that there will not be so many irrelevant changes? Remember also to modify proto's `srcDir` config in `build.gradle`

**Comment by birdstorm** [comment]
@disksing PTAL thx

File: src/main/java/org/tikv/common/region/RegionStoreClient.java [hunk 1]
```diff
@@ -427,13 +427,367 @@ private void handleRawBatchPut(RawBatchPutResponse resp) {
...
+      this.regionManager.onRequestFail(region);
+      throw new TiClientInternalException("CommitResponse failed without a cause");
+    }
+    if(resp.hasRegionError()){
+      //bo.doBackOff(BoRegionMiss, new RegionException(resp.getRegionError()));
+      //return false;
+      //Caller method should restart commit
+      throw new RegionException(resp.getRegionError());
+    }
+    //if hasLock, need to resolveLocks and retry?
```
Comment (disksing) [inline src/main/java/org/tikv/common/region/RegionStoreClient.java]: `Commit` should not meet `Locked` error.

File: src/main/java/org/tikv/common/region/RegionStoreClient.java [hunk 2]
```diff
@@ -427,13 +427,367 @@ private void handleRawBatchPut(RawBatchPutResponse resp) {
...
+                      this,
+                      region,
+                      resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+      Kvrpcpb.ScanLockResponse resp = callWithRetry(bo, TikvGrpc.METHOD_KV_SCAN_LOCK, factory, handler);
+      if (scanLockHelper(bo, resp)) {
+        return resp.getLocksList();
+      }
+
+      // we should refresh region
+      region = regionManager.getRegionByKey(startKey);
```
Comment (disksing) [inline src/main/java/org/tikv/common/region/RegionStoreClient.java]: why refresh region here?

File: src/main/java/org/tikv/txn/TikvTransaction.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Transaction implementation of TiKV client
+ */
+public class TikvTransaction implements ITransaction {
```
Comment (disksing) [inline src/main/java/org/tikv/txn/TikvTransaction.java]: Do we need `scan()`? If scan is needed, I think we will have to provide `scan` in `TikvTransaction` to merge keys in `memoryKv` and remote keys.

File: src/main/java/org/tikv/txn/TwoPhaseCommitter.java [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+            txnSize += (key.length + value.length);
+            //TODO check transaction maxEntrySize
+        }
+        List<byte[]> lockedKeys = transaction.getLockedKeys();
+        for(byte[] lockedKey : lockedKeys) {
+            Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
+                    .setKey(ByteString.copyFrom(lockedKey))
+                    .setOp(Kvrpcpb.Op.Lock)
+                    .build();
+            mutations.put(new String(lockedKey), mutation);
```
Comment (disksing) [inline src/main/java/org/tikv/txn/TwoPhaseCommitter.java]: If a key is already contained in `mutations`, we should not use the lock mutation to replace the old one.

File: src/main/java/org/tikv/txn/TwoPhaseCommitter.java [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        return this.doActionOnKeys(backOffer, TwoPhaseCommitType.actionCommit, keys);
+    }
+
+    /**
+     * 2pc - cleanup phase
+     * @param backOffer
+     * @param keys
+     * @return
+     */
+    public String cleanupKeys(BackOffer backOffer, byte[][] keys) {
```
Comment (disksing) [inline src/main/java/org/tikv/txn/TwoPhaseCommitter.java]: Is this method unused? (when a txn failed to commit, add keys that with prewrite request sent need to be cleaned up)

Comment (ujinccforever) [reply src/main/java/org/tikv/txn/TwoPhaseCommitter.java]: Thx for ur patient review & comment， we'll check these out then feedback as soon as possible ^_^

File: src/main/java/org/tikv/common/Snapshot.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  public ByteString get(ByteString key) {
+    RegionStoreClient client = clientBuilder.build(key);
+    // TODO: Need to deal with lock error after grpc stable
+    return client.get(ConcreteBackOffer.newGetBackOff(), key, timestamp.getVersion());
+  }
+
+  public Iterator<KvPair> scan(ByteString startKey) {
+    BackOffer bo = ConcreteBackOffer.newTsoBackOff();
+    long version = pdClient.getTimestamp(bo).getVersion();
+    return new ConcreteScanIterator(conf, clientBuilder, startKey, version);
```
[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/common/Snapshot.java]:
```suggestion
    BackOffer bo = ConcreteBackOffer.newTsoBackOff();
    return new ConcreteScanIterator(conf, clientBuilder, startKey, this.timestamp.getVersion());
```

[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/common/Snapshot.java]: Thx, this is a bit of a problem writing like this.

File: src/main/java/org/tikv/common/region/RegionStoreClient.java [hunk 7]
```diff
@@ -427,13 +427,367 @@ private void handleRawBatchPut(RawBatchPutResponse resp) {
...
+              region,
+              resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+      Kvrpcpb.DeleteRangeResponse resp = callWithRetry(backOffer, TikvGrpc.METHOD_KV_DELETE_RANGE, factory, handler);
+      if(deleteHelper(backOffer, resp)){
+        break;
+      }
+    }
+  }
+
+  private boolean deleteHelper(BackOffer bo, Kvrpcpb.DeleteRangeResponse resp) {
```
[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/common/region/RegionStoreClient.java]:
```suggestion
  private boolean deleteRangeHelper(BackOffer bo, Kvrpcpb.DeleteRangeResponse resp) {
```

[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/common/region/RegionStoreClient.java]: yeah right

File: src/main/java/org/tikv/common/region/RegionStoreClient.java (hunk 2 above)
Comment (jackycchen) [reply src/main/java/org/tikv/common/region/RegionStoreClient.java]: When encountering an region error, we should request for region info to PD.

File: src/main/java/org/tikv/txn/TwoPhaseCommitter.java (hunk 4 above)
Comment (jackycchen) [reply src/main/java/org/tikv/txn/TwoPhaseCommitter.java]:
```suggestion
            ByteString key = ByteString.copyFrom(lockedKey);
            if(this.mutations.get(key) != null) {
                //locked key exist already, omit it.
                continue;
            }
            Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
                    .setKey(key)
                    .setOp(Kvrpcpb.Op.Lock)
                    .build();
            mutations.put(new String(lockedKey), mutation);
```

File: src/main/java/org/tikv/txn/TwoPhaseCommitter.java (hunk 5 above)
Comment (jackycchen) [reply src/main/java/org/tikv/txn/TwoPhaseCommitter.java]: Yes, should add prewrite request when a txn encountered commit failed, I will fix it.

File: src/main/java/org/tikv/common/Snapshot.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.tikv.common.util.KeyRangeUtils.makeRange;
+
+public class Snapshot {
+  private final TiTimestamp timestamp;
+  //private final TiSession session;
```
Comment (AndreMouche) [inline src/main/java/org/tikv/common/Snapshot.java]: Please remove the code which is not used.

File: src/main/java/org/tikv/txn/TxnKVClient.java [hunk 9]
```diff
@@ -1,5 +1,363 @@
...
+     * Commit request of 2pc,
+     * add backoff logic when encountered region error, ErrBodyMissing, and other errors
+     * @param backOffer
+     * @param keys
+     * @param startTs
+     * @param commitTs
+     * @param regionId
+     * @return
+     */
+    public ClientRPCResult commit(BackOffer backOffer, byte[][] keys, long startTs, long commitTs, long regionId) {
```
Comment (AndreMouche) [inline src/main/java/org/tikv/txn/TxnKVClient.java]: I think as an SDK, the user needn't know 2PC,  so could we make functions `prewrite` and `commit` private?

File: src/main/java/org/tikv/txn/TxnKVClient.java [hunk 10]
```diff
@@ -1,5 +1,363 @@
...
+                        .build()
+        );
+        long lockTTL = 2000;
+        long startTS;
+        TiRegion region = regionManager.getRegionByKey(byteKey);
+        boolean prewrite = false;
+        while(true) {
+            try {
+                startTS = this.getTimestamp().getVersion();
+                ClientRPCResult prewriteResp = this.prewrite(bo, mutations, key, lockTTL, startTS, region.getId());
```
Comment (AndreMouche) [inline src/main/java/org/tikv/txn/TxnKVClient.java]: Does it seem we can only get a retriable error when meeting `RegionException`? So could we use the same `startTS` when we meet the `RegionException`?

File: src/main/java/org/tikv/txn/TxnKVClient.java [hunk 11]
```diff
@@ -1,5 +1,363 @@
...
+     */
+    public ITransaction begin() {
+        return new TikvTransaction(this);
+    }
+
+    public ITransaction begin(Function<ITransaction,Boolean> function) {
+        return new TikvTransaction(this, function);
+    }
+
+    //add backoff logic when encountered region error,ErrBodyMissing, and other errors
```
[the remark this answers is not in the document]
Comment (ujinccforever) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: We r just learning to refresh our coding style, thx for advice....

File: src/main/java/org/tikv/common/Snapshot.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2017 PingCAP, Inc.
```
[the remark this answers is not in the document]
Comment (ujinccforever) [reply src/main/java/org/tikv/common/Snapshot.java]: Exactly ^_^

File: src/main/java/org/tikv/txn/TxnKVClient.java (hunk 9 above)
Comment (ujinccforever) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: > I think as an SDK, the user needn't know 2PC, so could we make functions `prewrite` and `commit` private?

Maybe upper KV-engine driven encapsulation will need 2PC process to be customized...

File: src/main/java/org/tikv/txn/TxnKVClient.java [hunk 13]
```diff
@@ -1,5 +1,363 @@
...
+        boolean putResult = false;
+        ByteString byteKey = ByteString.copyFrom(key);
+        ByteString byteValue = ByteString.copyFrom(value);
+        BackOffer bo = ConcreteBackOffer.newCustomBackOff(BackOffer.prewriteMaxBackoff);
+        List<Kvrpcpb.Mutation> mutations = Lists.newArrayList(
+                Kvrpcpb.Mutation.newBuilder()
+                        .setKey(byteKey).setValue(byteValue).setOp(Kvrpcpb.Op.Put)
+                        .build()
+        );
+        long lockTTL = 2000;
```
[the remark this answers is not in the document]
Comment (ujinccforever) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: ORZ

[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: Get it

File: src/main/java/org/tikv/txn/TxnKVClient.java (hunk 10 above)
Comment (jackycchen) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: You means that only exception which contains Write and write conflict error should refresh txn startTs ? RegionException meets when server encounters region slipt , we needn't change startTs in this situation.
Is it right ?

Comment (jackycchen) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: > Does it seem we can only get a retriable error when meeting `RegionException`? So could we use the same `startTS` when we meet the `RegionException`?

You means that only exception which contains Write and write conflict error should refresh txn startTs ? RegionException meets when server encounters region slipt , we needn't change startTs in this situation.
Is it right ?

File: src/main/java/org/tikv/common/Snapshot.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jackycchen) [reply src/main/java/org/tikv/common/Snapshot.java]: Ok, copyright information will be modified nniformly.

File: src/main/java/org/tikv/txn/TikvTransaction.java (hunk 3 above)
Comment (ice1000) [reply src/main/java/org/tikv/txn/TikvTransaction.java]: ... And I believe we should reach an agreement on `TiKV`/`Tikv`.

**Comment by birdstorm** [comment]
@jackycchen Hi, any updates? :)

**Comment by ice1000** [comment]
Hi, please resolve master conflicts, thanks!

**Comment by jackycchen** [comment]
> Hi, please resolve master conflicts, thanks!

done

**Comment by jackycchen** [comment]
> @jackycchen Hi, any updates? :)

Updated

**Comment by AndreMouche** [comment]
Hi, Please address all comments @jackycchen

File: src/main/java/org/tikv/common/memdb/BufferStore.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            Key val = entry.getValue();
+            byte[] value = val.getBytes();
+            if(value.length == 0) {
+                boolean deleted = mutator.delete(key);
+                return deleted ? null : "delete [" + key.toString() + "] failed";
+            }
+            boolean inserted = mutator.set(key, value);
+            return inserted ? null : "set [" + key.toString() + "," + val.toString() + "] failed";
+        });
+        return false;
```
Comment (disksing) [inline src/main/java/org/tikv/common/memdb/BufferStore.java]: why return false here?

File: src/main/java/org/tikv/common/memdb/MemDbBuffer.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            //ErrTxnTooLarge
+            return false;
+        }
+        return success;
+    }
+
+    @Override
+    public boolean delete(Key key) {
+        try {
+            this.db.remove(key);
```
Comment (disksing) [inline src/main/java/org/tikv/common/memdb/MemDbBuffer.java]: I think you need to insert an entry to mark the key deleted. See https://github.com/pingcap/tidb/blob/538587a4a8e23b0207e1f1e3e867afb7bd3f8c9f/kv/memdb_buffer.go#L108-L111

File: src/main/java/org/tikv/common/memdb/MemDbBuffer.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        byte[] valueBytes = value != null && value.getBytes().length > 0 ? value.getBytes() : null;
+        return valueBytes;
+    }
+
+    @Override
+    public Iterator iterator(Key key, Key upperBound) {
+        if(key == null && upperBound == null) {
+            return this.db.entrySet().iterator();
+        }
+        if(key == null || upperBound == null) {
```
Comment (disksing) [inline src/main/java/org/tikv/common/memdb/MemDbBuffer.java]: In TiDB, `nil, nil` means iterate all keys.

Comment (disksing) [reply src/main/java/org/tikv/common/memdb/MemDbBuffer.java]: `key == nil` means starts from the first key. `upperBound == nil` means no upperbound.

File: src/main/java/org/tikv/common/memdb/BufferStore.java (hunk 14 above)
Comment (jackycchen) [reply src/main/java/org/tikv/common/memdb/BufferStore.java]: Thanks, it's a mistake here, will be fixed.

File: src/main/java/org/tikv/txn/type/BatchKeys.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import org.tikv.common.region.TiRegion;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class BatchKeys {
+    private List<byte[]> keys;
+
+    private Long regioId;
```
Comment (disksing) [inline src/main/java/org/tikv/txn/type/BatchKeys.java]: regionId

File: gradle/wrapper/gradle-wrapper.properties [hunk 18]
```diff
@@ -1,5 +1,6 @@
+#Wed Feb 13 17:14:42 CST 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-all.zip
```
[the remark this answers is not in the document]
Comment (jackycchen) [reply gradle/wrapper/gradle-wrapper.properties]: ok, you r right.

File: src/main/java/org/tikv/txn/type/BatchKeys.java (hunk 17 above)
Comment (jackycchen) [reply src/main/java/org/tikv/txn/type/BatchKeys.java]: > regionId

My mistake, thanks for the correction

File: src/main/java/org/tikv/common/Snapshot.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        try (RegionStoreClient client = clientBuilder.build(key)) {
+          List<KvPair> partialResult =
+              client.batchGet(backOffer, keyBuffer, timestamp.getVersion());
+          // TODO: Add lock check
+          result.addAll(partialResult);
+        } catch (Exception e) {
+          throw new TiClientInternalException("Error Closing Store client.", e);
+        }
+        keyBuffer = new ArrayList<>();
+        keyBuffer.add(key);
```
[the remark this answers is not in the document]
Comment (forget6) [reply src/main/java/org/tikv/common/Snapshot.java]: we didn't use the batchGet，we delete the method

File: src/main/java/org/tikv/txn/TxnKVClient.java [hunk 20]
```diff
@@ -1,5 +1,363 @@
...
+                       RegionStoreClient.RegionStoreClientBuilder clientBuilder, PDClient pdClient) {
+        this.conf = conf;
+        this.clientBuilder = clientBuilder;
+        this.regionManager = clientBuilder.getRegionManager();
+        //this.session = TiSession.create(TiConfiguration.createRawDefault(addresses));
+        //this.regionManager = new RegionManager(session.getPDClient());
+        this.pdClient = pdClient;
+    }
+
+    /*public static TxnKVClient createClient(String addresses) {
```
[the remark this answers is not in the document]
Comment (forget6) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: ok

File: src/main/java/org/tikv/txn/TxnKVClient.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (forget6) [reply src/main/java/org/tikv/txn/TxnKVClient.java]: ok

**Pull request closed without merging**